### PR TITLE
Add client for http4s library

### PR DESCRIPTION
**Pull request by igor-vovk**
Hi! I don't know if there is space for another client, but it seemed appropriate for me to add it, since it's kinda mainstream client in `cats-effect` ecosystem (http4s/circe/cats-effect). I was kinda surprised not to find it, ended up implementing it, but got tired copying it between my projects where I use elastic4s.

If you think it's a good idea to add it, I will be happy to add tests in this PR as well. Thank you.

### CODE REVIEW

**Comment by igor-vovk** [comment]
It just came to me, @Philippus if you think that `elastic4s` supports enough clients out of the box (I found that sttp for example already has http4s backend), I can just publish it as a separate library.

**Comment by lenguyenthanh** [comment]
I would love to use this client! I think currently sttp use future backend (which is backed scala future & akka iiuc). So this is definitely worth adding imho.

Also We probably can have more perform IO backed client if We have this: https://github.com/Philippus/elastic4s/issues/2379

**Comment by igor-vovk** [comment]
@lenguyenthanh I was also thinking about that, addressed in https://github.com/Philippus/elastic4s/pull/3119

**Comment by lenguyenthanh** [comment]
> @lenguyenthanh I was also thinking about that, addressed in #3119

this is great, thanks! I was about to do it.

**Comment by Philippus** [comment]
I think adding a new client is fine, provided someone is using it in production (like you). Maybe we can mark it as experimental, allowing us to make backwards incompatible changes for a certain time.

**Comment by igor-vovk** [comment]
@Philippus nice! Then I'm going to add tests for it

**Comment by igor-vovk** [comment]
@Philippus I've added some tests following the pattern from sttp. I think it is ready to be merged, WDYT?

**Comment by Philippus** [comment]
@igor-vovk now that https://github.com/Philippus/elastic4s/pull/3124 is merged, care to get back and clean up this PR (or create a new one based on the current main branch)?

**Comment by Philippus** [comment]
@lenguyenthanh Do you have time to take a look at the PR to see if this will work for you as well?

**Comment by lenguyenthanh** [comment]
> @lenguyenthanh Do you have time to take a look at the PR to see if this will work for you as well?

of course, I'll take a look tomorrow.

File: elastic4s-client-http4s/src/main/scala/com/sksamuel/elastic4s/http4s/Elastic4sEntityEncoders.scala
Comment (lenguyenthanh) [inline elastic4s-client-http4s/src/main/scala/com/sksamuel/elastic4s/http4s/Elastic4sEntityEncoders.scala:38]:
```suggestion
    new http4s.EntityEncoder[F, Option[A]] {
      override def toEntity(a: Option[A]): http4s.Entity[F] =
        a.fold[http4s.Entity[F]](http4s.Entity.empty)(ee.toEntity)
```
nit

File: elastic4s-client-http4s/src/main/scala/com/sksamuel/elastic4s/http4s/RequestResponseConverters.scala
Comment (lenguyenthanh) [inline elastic4s-client-http4s/src/main/scala/com/sksamuel/elastic4s/http4s/RequestResponseConverters.scala:19]:
```suggestion
      scheme = http4s.Uri.Scheme.fromString(endpoint.protocol).toOption,
```
nit

File: elastic4s-client-http4s/src/main/scala/com/sksamuel/elastic4s/http4s/RequestResponseConverters.scala
Comment (lenguyenthanh) [inline elastic4s-client-http4s/src/main/scala/com/sksamuel/elastic4s/http4s/RequestResponseConverters.scala:20]: as we already have `cats.syntax.all._` we can use `.some` as it nicer imho.

**Review by lenguyenthanh — APPROVED** [review]
I only have few nit comments, otherwise lgtm. And this will be so much better with the other of @igor-vovk's pr: https://github.com/Philippus/elastic4s/pull/3119

**Comment by igor-vovk** [comment]
@lenguyenthanh I've addressed all your comments

File: elastic4s-client-http4s/src/test/scala/com/sksamuel/elastic4s/http4s/Http4sRequestHttpClientTest.scala
Comment (Philippus) [inline elastic4s-client-http4s/src/test/scala/com/sksamuel/elastic4s/http4s/Http4sRequestHttpClientTest.scala:22]: Looks like some times the test gives back `yellow`, which I think for this test is ok.

**Review by Philippus — CHANGES_REQUESTED** [review]

File: elastic4s-client-http4s/src/test/scala/com/sksamuel/elastic4s/http4s/Http4sRequestHttpClientTest.scala
Comment (igor-vovk) [reply elastic4s-client-http4s/src/test/scala/com/sksamuel/elastic4s/http4s/Http4sRequestHttpClientTest.scala]: @Philippus fixed. I think it would be a good idea to extract a common test suite for all http clients (since they all can be tested in a single way), will try to do it in a separate PR, and then more tests can be added

**Comment by igor-vovk** [comment]
@Philippus I also noticed test flakiness, tried to add health check before proceeding on the level of GH actions in #3207, but failing to understand why it fails in GH Actions environment (works locally, probably something with bash options...)

**Comment by Philippus** [comment]
I think it's because of all the tests running on the same instance. If you change the test to "yellow" or "green" I think it will be fine.

**Comment by igor-vovk** [comment]
@Philippus I think what happens is that this status changes from `red` to `yellow` and then to `green` (doc https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-health.html). And then depending on when the test runs it can receive any of these, thus flakiness of the test. In any case, I've deleted this test. Should I bring it back?

**Comment by igor-vovk** [comment]
Marked #3207 as ready for review. It now waits elastic to start before continuing, should reduce tests flakiness

**Comment by Philippus** [comment]
> @Philippus I think what happens is that this status changes from `red` to `yellow` and then to `green` (doc https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-health.html). And then depending on when the test runs it can receive any of these, thus flakiness of the test. In any case, I've deleted this test. Should I bring it back?

If you think it's worth having it as a test I think we can bring it back. We can always remove it again, if it turns out to be flaky again.

**Comment by igor-vovk** [comment]
Reverted the deletion of the test. After #3207 will be merged, the test should stop being flaky

**Pull request merged**
